### DISCUSSION TITLE
Yolov8 Segmentation model output

### DISCUSSION THREAD

**Discussion by TalonBvV** [ultralytics/ultralytics#6623] [Q&A]
Hey there, I hope you are well.

I am having some trouble interpreting the output of my custom trained yolov8 model.

I trained the model on my own augmented version of the publaynet dataset.
I am using ncnn in C++ for inference.

I did find other people and a paper mentioning the post processing methods where you have to calculate the final mask based on the mask weights and prototype masks, but I haven't even gotten this far yet.

Could someone please assist me by providing more simplistic instructions, I just need to understand what variables are present in the output, and given the output shape: ((1, 42, 8400), (1, 32, 152, 200)) how do I locate them?

Sorry for the newbe question, this is my first time getting this far with a yolov8 segment model.

**Comment by pderrenger** [comment]
@TalonBvV hello! Congratulations on training your custom YOLOv8 segmentation model. 🎉

The output of a YOLOv8 segmentation model typically consists of two main components: detection outputs and segmentation outputs. The detection outputs include the bounding boxes, class confidences, and class predictions, while the segmentation outputs provide the per-pixel class predictions.

Given your output shape `((1, 42, 8400), (1, 32, 152, 200))`, it seems like you have two tensors:

1. The first tensor `(1, 42, 8400)` likely corresponds to the detection outputs, where `42` might represent the number of classes plus 5 (for x, y, width, height, and objectness score) times the number of anchors used, and `8400` is the flattened spatial grid of predictions.

2. The second tensor `(1, 32, 152, 200)` is probably the segmentation output, where `32` could be the number of prototype masks, and `(152, 200)` is the spatial resolution of the segmentation map.

To interpret these outputs:

- **Detection Outputs**: You would need to reshape the first tensor to the appropriate grid size and apply post-processing steps like thresholding based on confidence scores, non-maximum suppression (NMS), and possibly decoding the bounding box coordinates from the grid cell format to the image coordinate format.

- **Segmentation Outputs**: For the second tensor, you would typically apply a sigmoid or softmax function to convert the raw scores to probabilities and then select the class with the highest probability for each pixel to form the final segmentation mask.

Since you're using ncnn in C++, you'll need to implement the post-processing steps in your inference code. The exact details of these steps can be complex and depend on the specifics of your model and dataset.

For more detailed guidance on post-processing and interpreting the outputs, I recommend checking out the [Segmentation documentation](https://docs.ultralytics.com/tasks/segment). If you need further assistance, feel free to open an issue in the repo with specific details about your model and the outputs you're getting, and the community will be happy to help you out.

Keep up the great work, and don't hesitate to reach out if you have more questions! 🚀

**Reply by Holsonn** [reply]
> The first tensor (1, 42, 8400) likely corresponds to the detection outputs, where 42 might represent the number of classes plus 5 (for x, y, width, height, and objectness score) times the number of anchors used, and 8400 is the flattened spatial grid of predictions.

Hi! How to know the number of anchors used?

**Reply by pderrenger** [reply]
Hello! 😊 The number of anchors used in your model can typically be found in the model's configuration file (e.g., `yolov8n.yaml`). Look for a section named `anchors` or similar, where the anchor sizes are defined. Each row in this section represents a different scale of anchors, and the number of rows gives you the number of anchor groups used.

For example:
```yaml
anchors:
  - [10,13, 16,30, 33,23]  # Anchors for small objects
  - [30,61, 62,45, 59,119]  # Anchors for medium objects
  - [116,90, 156,198, 373,326]  # Anchors for large objects
```
In this case, there are 3 groups of anchors used. If your model configuration differs or if you're dynamically generating anchors, the details should be documented in the model's configuration or creation script.

Hope this helps! Let us know if you have any more questions.
